### PR TITLE
l10n_ch_hr_payroll: Improvement of salary rules and employee datas 

### PR DESCRIPTION
**Pull request by open-net-sarl**
V1.0.3-4: 
- Add LPP Amount to contract. 
- Add Worked Hours Rate to contract. 
- Add Hourly Rate to contract. 
- Compute the Wage of contract, based on Worked Hours and Hourly Rate. 
- Add new salay rules  

V1.0.5: 
- Add Holiday Rate to contract. 
- Update "Indemnité vacances 8,33%" rule to include "Holiday Rate" from contract.  

V1.0.6: 
- Move salary rules from CSV file to XML file. 
- Import salary rules only at install. 
- Internal reorganization of files (.py and .xml).  

Attention: To prevent overwrite your salary rules changes, an update from 1.0.5 and lower to 1.0.6 and higher create duplicates of salary rule

### CODE REVIEW

File: l10n_ch_hr_payroll/README.rst [hunk 1]
```diff
@@ -19,13 +24,29 @@ info@open-net.ch
...
+    * Add new salay rules
+
+V1.0.5:
+    * Add Holiday Rate to contract.
+    * Update "Indemnité vacances 8,33%" rule to include "Holiday Rate" from contract.
+
+V1.0.6:
+    * Move salary.rule from CSV file to XML file.
+    * Import salary.rule only at install.
+    * Internal reorganization of files (.py and .xml).
```
Comment (yvaucher) [inline l10n_ch_hr_payroll/README.rst:58]: Please take the opportunity to follow OCA README template:

https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

**Comment by nbessi** [comment]
@gurneyalex can you have a look for user @Sge-opennet. I think they have a company CLA

**Comment by gurneyalex** [comment]
they do. database updated, CLA ok.

**Comment by open-net-sarl** [comment]
Hello,

Could you process the merge please ?

**Comment by open-net-sarl** [comment]
Hello,

Thanks for merging

**Comment by gurneyalex** [comment]
runbot warnings unrelated to this PR.

**Comment by gurneyalex** [comment]
:+1: although automated tests would be nice.

File: l10n_ch_hr_payroll/README.rst [hunk 2]
```diff
@@ -1,31 +1,73 @@
...
-=========================
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
 
-**Features list :**
-    * Add Swiss salary rule categories
-    * Add Swiss salary rules
-    * Add children in school to employee
-    * Add LPP range to contract
+Module name
```
Comment (yvaucher) [inline l10n_ch_hr_payroll/README.rst]: Please put `Switzerland Payroll Rules` instead of "Module name"

**Comment by open-net-sarl** [comment]
When I try to access to the result of ci/runbot check, the website return an error 500.
I've got this URL: https://runbot.odoo-community.org/runbot/build/2946996

File: l10n_ch_hr_payroll/README.rst [hunk 3]
```diff
@@ -1,31 +1,73 @@
...
+
+Contributors
+------------
+
+* Sebastien Legendre <sge@open-net.ch>
+* Yvon-Philippe Crittin <cyp@open-net.ch>
+
+Maintainer
+----------
+
```
Comment (yvaucher) [inline l10n_ch_hr_payroll/README.rst:77]: Here maintainer should be Odoo community association as in template.

We want issues reported on github to track them as we do for every single module included in the OCA.

File: l10n_ch_hr_payroll/__openerp__.py [hunk 4]
```diff
@@ -26,23 +26,42 @@
 #
 ##############################################################################
 
+
 {
     'name': 'Switzerland - Payroll',
     'summary': 'Swizerland Payroll Rules',
     'category': 'Localization',
     'author': "Open-Net Sàrl,Odoo Community Association (OCA)",
+    'description': """
```
Comment (yvaucher) [inline l10n_ch_hr_payroll/__openerp__.py]: Please remove description from `__openerp__.py` moving feature list in `README.rst`

`README.rst` will be displayed in Odoo and on apps.odoo.com only if `description` key is not present in `__openerp__.py`

File: l10n_ch_hr_payroll/data/hr.salary.rule.xml [hunk 5]
```diff
@@ -0,0 +1,531 @@
...
+    <!-- </data> -->
+    <!-- <data> -->
+        <record id="hr_payroll.hr_rule_taxable" model="hr.salary.rule">
+            <field name="amount_select">code</field>
+            <field name="appears_on_payslip">True</field>
+            <field name="code">GROSS</field>
+            <field name="condition_select">none</field>
+            <field name="name">Gross</field>
+            <field name="amount_python_compute">result = categories.BASIC + categories.ALW + categories.FALW</field>
+            <field name="condition_python"> # Available variables: #---------------------- # payslip: object containing the payslips # employee: hr.employee object # contract: hr.contract object # rules: object containing the rules code (previously computed) # categories: object containing the computed salary rule categories (sum of amount of all rules belonging to that category). # worked_days: object containing the computed worked days # inputs: object containing the computed inputs  # Note: returned value have to be set in the variable 'result'  result = rules.NET > categories.NET * 0.10</field>
```
Comment (yvaucher) [inline l10n_ch_hr_payroll/data/hr.salary.rule.xml]: Shouldn't you add line break of some kind in this line. Isn't this a one line string ?

Tested on runbot, this is shown as a single line if you change "Condition Based On" to "Python Expression" (`python` selection value) (shouldn't it be that by default for field `condition_select`?)

File: l10n_ch_hr_payroll/data/hr.salary.rule.xml [hunk 6]
```diff
@@ -0,0 +1,531 @@
...
+            <field name="category_id" ref="hr_payroll.GROSS"/>
+        </record>
+        <record id="hr_payroll.hr_rule_net" model="hr.salary.rule">
+            <field name="amount_select">code</field>
+            <field name="appears_on_payslip">True</field>
+            <field name="code">NET</field>
+            <field name="condition_select">none</field>
+            <field name="name">Net</field>
+            <field name="amount_python_compute">result = categories.BASIC + categories.ALW + categories.FALW + categories.DED</field>
+            <field name="condition_python"> # Available variables: #---------------------- # payslip: object containing the payslips # employee: hr.employee object # contract: hr.contract object # rules: object containing the rules code (previously computed) # categories: object containing the computed salary rule categories (sum of amount of all rules belonging to that category). # worked_days: object containing the computed worked days # inputs: object containing the computed inputs  # Note: returned value have to be set in the variable 'result'  result = rules.NET > categories.NET * 0.10</field>
```
Comment (yvaucher) [inline l10n_ch_hr_payroll/data/hr.salary.rule.xml]: Same question here

File: l10n_ch_hr_payroll/README.rst [hunk 7]
```diff
@@ -1,31 +1,73 @@
-Swizerland Payroll Rules.
-=========================
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
 
-**Features list :**
```
Comment (yvaucher) [inline l10n_ch_hr_payroll/README.rst:4]: Please restore Feature list in here, this README.rst will be displayed as description of the module on Github, on Odoo instance and on apps.odoo.com

File: l10n_ch_hr_payroll/models/hr_contract.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
+# -*- coding: utf-8 -*-
+#
+#  File: hr_contract.py
+#  Module: l10n_ch_hr_payroll
+#
+#  Created by sge@open-net.ch
+#
+#  Copyright (c) 2014 Open-Net Ltd.
```
Comment (yvaucher) [inline l10n_ch_hr_payroll/models/hr_contract.py]: Don't you want to update the copyright date to 2014-2015 ?

File: l10n_ch_hr_payroll/data/hr.salary.rule-new.csv [hunk 9]
```diff
@@ -1,28 +0,0 @@
...
-30,Python Code,True,P13,Always True,,,,,Provision 13ème,,,result = ((categories.GROSS - categories.DED) / 12 ) + categories.COSTS, # Available variables: #---------------------- # payslip: object containing the payslips # employee: hr.employee object # contract: hr.contract object # rules: object containing the rules code (previously computed) # categories: object containing the computed salary rule categories (sum of amount of all rules belonging to that category). # worked_days: object containing the computed worked days # inputs: object containing the computed inputs  # Note: returned value have to be set in the variable 'result'  result = rules.NET > categories.NET * 0.10,1,contract.wage,250,Provision 13th salary,
```
Comment (guewen) [inline l10n_ch_hr_payroll/data/hr.salary.rule-new.csv:28]: After this file is removed, an update of the module on an existing instance will delete all the `hr.salary.rule` above.
Maybe you should keep the same xml ids in the xml file or provide a migration script that remap the xmlids in ir_model_data?

**Comment by yvaucher** [comment]
@open-net-sarl ~~Runbot is not fully functional, yet. So no need to worry about it being red.~~

You should now be able to connect to runbot

File: l10n_ch_hr_payroll/data/hr.salary.rule.xml (hunk 5 above)
Comment (yvaucher) [inline l10n_ch_hr_payroll/data/hr.salary.rule.xml]: To fix it simply remove this value.

AFAICS it is just a copy of default value setted here:
https://github.com/odoo/odoo/blob/8.0/addons/hr_payroll/hr_payroll.py#L797-811

File: l10n_ch_hr_payroll/data/hr.salary.rule.xml (hunk 6 above)
Comment (yvaucher) [inline l10n_ch_hr_payroll/data/hr.salary.rule.xml]: Remove this line

**Comment by open-net-sarl** [comment]
Still no access to the return of the cl/runbot: https://runbot.odoo-community.org/runbot/build/2947507

Error: HTTP 500 - Internal Server Error

:(

**Comment by open-net-sarl** [comment]
Seems the runbot is the source of errors, not the module

File: l10n_ch_hr_payroll/models/hr_employee.py [hunk 10]
```diff
@@ -5,7 +5,7 @@
 #
 #  Created by sge@open-net.ch
 #
-#  Copyright (c) 2014 Open-Net Ltd.
+#  Copyright (c) 2015 Open-Net Ltd.
```
Comment (yvaucher) [inline l10n_ch_hr_payroll/models/hr_employee.py]: You can't do that, copyright should keep creation year of the file.
2014-2015 would be right if you changed that file this year.

Or was this file created in 2015?

File: l10n_ch_hr_payroll/README.rst [hunk 11]
```diff
@@ -1,31 +1,94 @@
...
+.. image:: http://open-net.ch/logo.png
+   :alt: Open Net Sarl
+   :target: http://open-net.ch
+
+This module is maintained by Open Net Sarl.
+Industrie 59 
+CH-1030 Bussigny 
+
+http://www.open-net.ch
+info@open-net.ch
```
Comment (gurneyalex) [inline l10n_ch_hr_payroll/README.rst]: If you're submitting this to the OCA, it means that the module is maintained collectively by the OCA. Open Net is a contributor. Please remove the paragraph.

**Comment by open-net-sarl** [comment]
Concerning "ci/runbot": can someone explain the reasons for the two warnings:

openerp.models:4071 create ir.values.create() with unknown fields: object
openerp.modules.loading:372 load_modules The transient model report (Report) should not have explicit access rules!

Our module does not use the word "object", nor does it have any report.
Thanks in advance.

**Comment by open-net-sarl** [comment]
Can someone explain the two warning of the "ci/runbot"?

openerp.models:4071 create ir.values.create() with unknown fields: object
openerp.modules.loading:372 load_modules The transient model report (Report) should not have explicit access rules!

Our module does not use the word "object", nor does it have any report.
Thanks in advance.

**Comment by yvaucher** [comment]
@open-net-sarl For the second one, a module has an access rule on a report. To fix it, it is needed to remove it. As far as I know access rule on report have no effect

The first one seems to be comming from l10n_ch_sepa

So both warning are unrelated. Just ignore them. I'll make a PR to fix the one in l10n_ch_sepa

Only @guewen comment to answer/agree/disagree and we are good :+1: from me already

**Comment by open-net-sarl** [comment]
@yvaucher I have checked once again: our files does not include any access rule. Is it possible to have more details on the second warning?

**Comment by open-net-sarl** [comment]
@yvaucher I have checked once again: our module does not include anything that concerns an access rule. I wonder if it's possible to have more details about that warning, as it happens in the fourth stage (cleaning part of the loading process), which means that it may happen with any active module.

**Comment by guewen** [comment]
> @yvaucher I have checked once again: our module does not include anything that concerns an access rule. I wonder if it's possible to have more details about that warning, as it happens in the fourth stage (cleaning part of the loading process), which means that it may happen with any active module.

You are right. It is probably in another module.

**Comment by guewen** [comment]
Did you see this issue https://github.com/OCA/l10n-switzerland/pull/132#discussion_r31026436?

**Comment by nbessi** [comment]
@open-net-sarl @yvaucher  we were pointed by @gurneyalex that there where some deprecated security rule on wizard and report. Fixing it is in my pipe. So the red runbot should not be an issue

**Comment by yvaucher** [comment]
@open-net-sarl @guewen Indeed it's an other module

As I said: "So both warning are unrelated. Just ignore them. I'll make a PR to fix the one in l10n_ch_sepa"

Which one is here https://github.com/OCA/l10n-switzerland/pull/140

**Comment by open-net-sarl** [comment]
This branch has failed checks, but can be merged. 

Could you please do the merge ?

**Comment by yvaucher** [comment]
@open-net-sarl 3rd reminder:

You may have missed it but you still didn't gave an answer to this comment https://github.com/OCA/l10n-switzerland/pull/132#discussion_r31026436 about `l10n_ch_hr_payroll/data/hr.salary.rule-new.csv`
It is the last blocking point there.

**Comment by open-net-sarl** [comment]
It has been added in the README.rst

https://github.com/open-net-sarl/l10n-switzerland/commit/9b15fbd91aca72b0429adf5b2d33daa0e5be1cea

**Comment by guewen** [comment]
@open-net-sarl odoo allows to run migration scripts automatically when updating modules so that we don't need to run sql queries manually. Do you need some guidance to achieve that?

**Comment by guewen** [comment]
Shouldn't the new rules replace the old ones? If the old rules must be kept for the historic record, so they should still be in the csv file but with the active flag set to False.
If the new rules replace the old ones, the xmlids should be remapped on the new ones.

**Comment by yvaucher** [comment]
@open-net-sarl as @guewen pointed it out it would be better with a migration script.

Upgrading this module would otherwise create duplicate rules without hint for a simple user to know easily which one is the one to use and which one is the old one.

So if it is possible, it would be more user friendly to replace old rules by new ones.

Here are few examples of migration script:
https://github.com/OCA/l10n-switzerland/tree/8.0/l10n_ch_bank/migrations/8.0.9.0.0
https://github.com/OCA/l10n-canada/tree/8.0/res_partner_attributes_add_BN/migrations/8.0.1.0
https://github.com/OCA/purchase-workflow/tree/8.0/framework_agreement/migrations/8.0.2.0

**Comment by open-net-sarl** [comment]
Can someone explain me if there is something to do now?
Thanks in advance

File: l10n_ch_hr_payroll/migrations/8.0.1.0.09/pre-update-ir_model_data.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+
+def migrate(cr, version):
+    if not version:
+        return
+
+    query = ("UPDATE ir_model_data "
+             "SET noupdate=true "
+             "WHERE module='l10n_ch_hr_payroll' "
+             "AND model='hr.salary.rule' ")
```
Comment (yvaucher) [inline l10n_ch_hr_payroll/migrations/8.0.1.0.09/pre-update-ir_model_data.py]: Don't you want to map the new rules to the old ones? Won't you have duplicate rules?

**Comment by open-net-sarl** [comment]
Hi,
The rules should have been 'non updatable' from version 1.0
The rules are 'models' but each customer is expected to change most of them, not AVS, but other salary rules are customer spécific.

So, if you installed version 1.0, you probably have adjusted many of the rules and don't want the update to change it. So we set them to 'non updatable' to protect the customer changes.

If you start from this version, you will have the new rules, directly set as non updatable

**Comment by open-net-sarl** [comment]
Hello,

We have other changes and improvement to publish for this module.

Could someone merge this pull ?

**Comment by guewen** [comment]
> The rules should have been 'non updatable' from version 1.0
> The rules are 'models' but each customer is expected to change most of them, not AVS, but other salary rules are customer spécific.
> 
> So, if you installed version 1.0, you probably have adjusted many of the rules and don't want the update to change it. So we set them to 'non updatable' to protect the customer changes.
> 
> If you start from this version, you will have the new rules, directly set as non updatable

It makes sense. Thanks for the migration script. There is just a last little issue that I'll add as inline comment.

File: l10n_ch_hr_payroll/migrations/8.0.1.0.09/pre-update-ir_model_data.py (hunk 12 above)
Comment (yvaucher) [inline l10n_ch_hr_payroll/migrations/8.0.1.0.09/pre-update-ir_model_data.py]: @open-net-sarl Ok thanks for the answer.

File: l10n_ch_hr_payroll/migrations/8.0.1.0.09/pre-update-ir_model_data.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
```
Comment (guewen) [inline l10n_ch_hr_payroll/migrations/8.0.1.0.09/pre-update-ir_model_data.py]: As the version in `__openerp__.py` is `1.0.8`, the folder should be `8.0.1.0.8` and not `8.0.1.0.09` (as well as the header's comment). Now the migration script is not loaded. Could you correct that?

**Comment by guewen** [comment]
Squashing some commits with `git rebase -i` would be welcome

**Comment by open-net-sarl** [comment]
@guewen I just did this, but do I need to add something to the command line?
Because I have done this locally, but it doesn't seem to influence the flow here.
For example, do I need to add the branch, same as when I push any modification with 
"git push ons_github flex_hours_updt"
Sorry to ask but I am new to git.

**Comment by guewen** [comment]
After a `git rebase -i`, the history is modified so you need to force the push with `git push -f`. No problem with asking!

**Comment by guewen** [comment]
Sorry for the version number in the file it is not correct yet. You wrote `8.0.1.0.08` when it should be `8.0.1.0.8`. Why do you add a `0` in there?

**Comment by open-net-sarl** [comment]
@guewen Sorry about the leading 0: it is a kind of reflex for sorting out the strings. I'm going to remove it.

**Comment by open-net-sarl** [comment]
@guewen  It seems it's all fine as the Travis fail is unrelated to this module

**Comment by guewen** [comment]
@open-net-sarl Thanks for the changes. Have had you a problem with the `git rebase -i`? The goal is to use "squash" on commits with similar changes. Do you need help?

**Comment by guewen** [comment]
Once the commits are cleaned you'll have my thumb up.

**Comment by open-net-sarl** [comment]
@guewen one last push, just after the git rebase -i, to be sure. No problems happened.
Thank you for your help.

**Comment by open-net-sarl** [comment]
Hi,

Is your thumb up ?

**Comment by guewen** [comment]
Seems there is a misunderstanding on the `git rebase -i` usage. See here https://github.com/OCA/l10n-switzerland/pull/132/commits, there are a lot of commits, often with the same commit message.
When you proceed with `git rebase -i`, what you should do is to write `f` or `s` (for fixup or squash) instead of `keep` before the commits and perhaps to move some of them before in order to join several commits together (the ones doing the same type of change). See https://fr.atlassian.com/git/tutorials/rewriting-history/git-rebase-i

Here for instance, you can fixup 3291b60, 426fed2, 86eb558 and 4943ea4 in fe9fee3 because they are the commits that set up the migration script, and so on with the other commits.

**Comment by open-net-sarl** [comment]
@guewen  Sorry but it didn't worked. First: thanks for the link to the tutorial: really interesting. But there is something that I have probably missed...
I have launched the command: git rebase -i flex_hours_updt
Then in the editor I put
pick fe9fee3
fixup 4943ea4
fixup 86eb558
fixup 426fed2
fixup 3291b60

Upon exiting, the response has been "could not apply fe9fee3".
Once again, thanks for your help.

**Comment by guewen** [comment]
Sometimes you may have to resolve a conflict and continue with `git rebase --continue`

**Comment by open-net-sarl** [comment]
@guewen thanks for your help. The last "rebase" have been done, and validated by the system.
What should we do, now?

**Comment by guewen** [comment]
You didn't `git rebase -i` anything actually. I still see 64 commits where I expect to see much less of them. Example: "`HaHa.`" https://github.com/open-net-sarl/l10n-switzerland/commit/eda74a3b24e68bfa8e731a82ae915ec892b77f01 is not a commit message that we should see in the history, to be squashed in another one or at least reworded. Did you forget to push your rebased branch?

**Comment by guewen** [comment]
I am a bit uneasy because I would like this to be merged too so we can go on, what I ask seems a big deal to you and that's not comfortable for any of us. Though, I really think that you should learn how to use this `git rebase -i` thing because it will be asked to you more than once. So I am disposed to help you again if you need more tips.

**Comment by open-net-sarl** [comment]
@guewen I'm really sorry for the mess and to make you feel uneasy. There are two problems: I have not started this branch, and I would never have published dummy comments such as "Haha.".
Furthermore, I have tried to follow the tutorial you indicated but it's an evidence: the result is not good.
In the end, the best solution for now would be to throw away everything  in "flex_hours_updt", and to re-commit only one version: the actual, last one (I have a local copy).
Reason: nothing interesting in-between, no need to keep whatever it is.
I have to find a solution for this.

**Comment by guewen** [comment]
@open-net-sarl This could be simpler indeed. This should allow you to do that :

```
git reset --soft 39007b0^
git commit -m "Your commit message"
git push -f
```

**Comment by open-net-sarl** [comment]
@guewen 
We merge them all in a single commit. Hope it's fine now

**Comment by guewen** [comment]
Did you do it with the commands I proposed you? Because actually there are a lot more things in the diff

**Comment by guewen** [comment]
You can correct your diff with the commands:

```
git rebase -s recursive -X ours origin/8.0 
git push -f
```

**Comment by open-net-sarl** [comment]
Indeed: that's what I did.. What do you mean by "ours origin/8.0" ?

**Comment by guewen** [comment]
It is the command you have to run. It means that it will proceed with a rebase towards `origin/8.0` where `origin` is the OCA branch. `-X ours` is there for helping git to resolve the conflicts, we decide that when it finds a conflict, it takes the remote (so OCA) code (`ours` and `theirs` are reversed when doing a `rebase`), it is valid here because the conflicts are in other files than your module.

**Comment by guewen** [comment]
(the other way being to run `git rebase origin/8.0` and fixing the conflicts one by one...)

**Comment by open-net-sarl** [comment]
Done. Once again: thank you for your explanations.

**Comment by guewen** [comment]
Much better! Thanks

:+1:  for merging

**Comment by yvaucher** [comment]
damn a last remaining tiny thing which makes the test red:

```
./l10n_ch_hr_payroll/models/hr_contract.py:64:1: W391 blank line at end of file
```

After that :+1: and we can certainly merge.
Thanks for the perseverance

**Comment by open-net-sarl** [comment]
... And my thanks for the patience...
:+1:

**Comment by guewen** [comment]
:grinning:

**Comment by open-net-sarl** [comment]
@guewen  Sorry to ask but I need to be sure. If I have to change anything in "l10n_ch_hr_payroll", the normal way here is to fork and then to ask for a merge, that's correct?.
Even if it concerns only the README.rst?
Thanks in advance

**Comment by guewen** [comment]
@open-net-sarl yes creating a new branch on your fork and open a pull request is the way even for trivial changes, which hopefully get merged faster usually

**Comment by open-net-sarl** [comment]
@guewen thank  you so much :+1:

**Pull request merged**